src/class/proto_methods.rs

```diff
@@ -0,0 +1,152 @@
+use crate::class::{
+    basic::PyObjectMethods, descr::PyDescrMethods, gc::PyGCMethods, iter::PyIterMethods,
+};
+use crate::ffi::{
+    PyAsyncMethods, PyBufferProcs, PyMappingMethods, PyNumberMethods, PySequenceMethods,
+};
+use std::{
+    ptr::{self, NonNull},
+    sync::atomic::{AtomicPtr, Ordering},
+};
+
+/// Defines what we need for method protocols.
+/// Stub implementations are for rust-numpy.
+pub trait PyProtoMethods {
+    fn async_methods() -> Option<NonNull<PyAsyncMethods>> {
+        None
+    }
+    fn basic_methods() -> Option<NonNull<PyObjectMethods>> {
+        None
+    }
+    fn buffer_methods() -> Option<NonNull<PyBufferProcs>> {
+        None
+    }
+    fn descr_methods() -> Option<NonNull<PyDescrMethods>> {
+        None
+    }
+    fn gc_methods() -> Option<NonNull<PyGCMethods>> {
+        None
+    }
+    fn mapping_methods() -> Option<NonNull<PyMappingMethods>> {
+        None
+    }
+    fn number_methods() -> Option<NonNull<PyNumberMethods>> {
+        None
+    }
+    fn iter_methods() -> Option<NonNull<PyIterMethods>> {
+        None
+    }
+    fn sequence_methods() -> Option<NonNull<PySequenceMethods>> {
+        None
+    }
+}
+
+/// Indicates that a type has a protocol registory.
+#[doc(hidden)]
+pub trait HasProtoRegistry: Sized + 'static {
+    fn registory() -> &'static PyProtoRegistry;
+}
+
+impl<T: HasProtoRegistry> PyProtoMethods for T {
+    fn async_methods() -> Option<NonNull<PyAsyncMethods>> {
+        NonNull::new(Self::registory().async_methods.load(Ordering::SeqCst))
+    }
+    fn basic_methods() -> Option<NonNull<PyObjectMethods>> {
+        NonNull::new(Self::registory().basic_methods.load(Ordering::SeqCst))
+    }
+    fn buffer_methods() -> Option<NonNull<PyBufferProcs>> {
+        NonNull::new(Self::registory().buffer_methods.load(Ordering::SeqCst))
+    }
+    fn descr_methods() -> Option<NonNull<PyDescrMethods>> {
+        NonNull::new(Self::registory().descr_methods.load(Ordering::SeqCst))
+    }
+    fn gc_methods() -> Option<NonNull<PyGCMethods>> {
+        NonNull::new(Self::registory().gc_methods.load(Ordering::SeqCst))
+    }
+    fn mapping_methods() -> Option<NonNull<PyMappingMethods>> {
+        NonNull::new(Self::registory().mapping_methods.load(Ordering::SeqCst))
+    }
+    fn number_methods() -> Option<NonNull<PyNumberMethods>> {
+        NonNull::new(Self::registory().number_methods.load(Ordering::SeqCst))
+    }
+    fn iter_methods() -> Option<NonNull<PyIterMethods>> {
+        NonNull::new(Self::registory().iter_methods.load(Ordering::SeqCst))
+    }
+    fn sequence_methods() -> Option<NonNull<PySequenceMethods>> {
+        NonNull::new(Self::registory().sequence_methods.load(Ordering::SeqCst))
+    }
+}
+
+#[doc(hidden)]
+pub struct PyProtoRegistry {
+    /// Async protocols.
+    async_methods: AtomicPtr<PyAsyncMethods>,
```

Comment:
Just because they are static but mutable.
Another solution(e.g.,  `RefCell<Option<NonNull<Methods>>>`, `fn registry() -> &'static RefCell<PyProtoRegistry>`)is also OK, but I'm not sure there's merit to do so.